src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs

```diff
@@ -0,0 +1,291 @@
+                // either the synchronous invocation of the callback completed or that the callback
+                // will never be invoked.
+                CancellationRegistration.Dispose();
+
+                // Now that we know the callback has quiesced, check to see whether it scheduled asynchronous
+                // work.  If it did, wait for that work to complete.
+                WorkItem.CallbackCompleted?.GetAwaiter().GetResult();
+            }
+
+            /// <summary>Asynchronously waits for any pending cancellation callback to complete and cleans up resources.</summary>
+            public async ValueTask DisposeAsync()
+            {
+                if (WorkItem is null)
+                {
+                    return;
+                }
+
+                // Prior to calling Dispose on the CancellationTokenRegistration, we need to tell
+                // the registration callback to exit if it's currently running; otherwise, we could deadlock.
+                WorkItem.ContinueTryingToCancel = false;
+
+                // Then we need to dispose of the registration.  Upon Dispose returning, we know that
+                // either the synchronous invocation of the callback completed or that the callback
+                // will never be invoked.
+                await CancellationRegistration.DisposeAsync().ConfigureAwait(false);
+
+                // Now that we know the callback has quiesced, check to see whether it scheduled asynchronous
+                // work.  If it did, wait for that work to complete.
+                if (WorkItem.CallbackCompleted is Task t)
+                {
+                    await t.ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <summary>Registers for cancellation with the specified token.</summary>
+        /// <remarks>Upon cancellation being requested, the implementation will attempt to CancelSynchronousIo for the thread calling RegisterCancellation.</remarks>
+        private SyncAsyncWorkItemRegistration RegisterCancellation(CancellationToken cancellationToken)
+        {
+            // If the token can't be canceled, there's nothing to register.
+            if (!cancellationToken.CanBeCanceled)
+            {
+                return default;
+            }
+
+            // Get a handle for the current thread. This is stored and used to cancel the I/O on this thread
+            // in response to the cancellation token having cancellation requested.  If the handle is invalid,
+            // which could happen if OpenThread fails, skip attempts at cancellation. The handle needs to be
+            // opened with THREAD_TERMINATE in order to be able to call CancelSynchronousIo.
+            ThreadHandle = t_currentThreadHandle ??= Interop.Kernel32.OpenThread(Interop.Kernel32.THREAD_TERMINATE, bInheritHandle: false, Interop.Kernel32.GetCurrentThreadId());
+            if (ThreadHandle.IsInvalid)
+            {
```

Comment:
I wonder in what circumstances this can be `true`? The [docs](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-openthread) don't help much:

> If the function fails, the return value is NULL. To get extended error information, call [GetLastError](https://docs.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).

If in the future some of the cancellation tests start to time out, we might consider adding some kind of `Debug.Fail($"OpenThread failed with {GetLastError()}")` here to see if we ever fail to "register".

Reply:
The only other error I've seen is for invalid arguments, e.g. if the thread ID doesn't exist.  I'll add an assert and see what happens :)